From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: WilyWork <hezodions@gmail.com>
Date: Wed, 1 Jul 2026 20:16:55 -0300
Subject: [PATCH] Add growable plant Block API

Introduce PandaGrowablePlant for age-based NMS plants and expose the data through the public Block API implementation.

This keeps the API-facing CraftBlock bridge and the underlying server abstraction in a single patch.

diff --git a/src/main/java/net/minecraft/server/BlockCactus.java b/src/main/java/net/minecraft/server/BlockCactus.java
index 85d89f2a279e3d0eb0f4da14240d5ebd3236ff5b..3f38733d1bb0dedfaf2d4ab0988ce70c6a9dc5a3 100644
--- a/src/main/java/net/minecraft/server/BlockCactus.java
+++ b/src/main/java/net/minecraft/server/BlockCactus.java
@@ -6,7 +6,7 @@ import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
 
 
 
-public class BlockCactus extends Block {
+public class BlockCactus extends Block implements PandaGrowablePlant { // PandaSpigot
 
     public static final BlockStateInteger AGE = BlockStateInteger.of("age", 0, 15);
 
@@ -117,6 +117,12 @@ public class BlockCactus extends Block {
         return ((Integer) iblockdata.get(BlockCactus.AGE)).intValue();
     }
 
+    // PandaSpigot start - expose shared plant growth state
+    public BlockStateInteger getGrowthState() {
+        return BlockCactus.AGE;
+    }
+    // PandaSpigot end
+
     protected BlockStateList getStateList() {
         return new BlockStateList(this, new IBlockState[] { BlockCactus.AGE});
     }
diff --git a/src/main/java/net/minecraft/server/BlockCocoa.java b/src/main/java/net/minecraft/server/BlockCocoa.java
index 206ca5c1a23d13d19e913eb9393dd09678e0aed3..ca82eac7157db18ce4526f5440ad3eb0c5ee73fa 100644
--- a/src/main/java/net/minecraft/server/BlockCocoa.java
+++ b/src/main/java/net/minecraft/server/BlockCocoa.java
@@ -5,7 +5,7 @@ import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
 
 
 
-public class BlockCocoa extends BlockDirectional implements IBlockFragilePlantElement {
+public class BlockCocoa extends BlockDirectional implements IBlockFragilePlantElement, PandaGrowablePlant { // PandaSpigot
 
     public static final BlockStateInteger AGE = BlockStateInteger.of("age", 0, 2);
 
@@ -155,6 +155,12 @@ public class BlockCocoa extends BlockDirectional implements IBlockFragilePlantEl
         return i;
     }
 
+    // PandaSpigot start - expose shared plant growth state
+    public BlockStateInteger getGrowthState() {
+        return BlockCocoa.AGE;
+    }
+    // PandaSpigot end
+
     protected BlockStateList getStateList() {
         return new BlockStateList(this, new IBlockState[] { BlockCocoa.FACING, BlockCocoa.AGE});
     }
diff --git a/src/main/java/net/minecraft/server/BlockCrops.java b/src/main/java/net/minecraft/server/BlockCrops.java
index 23149c658a1c7116fb9fb021a4df5c9127ada5b3..48baaccd0e47b8d4ea93ba554ea7f45b7faba89a 100644
--- a/src/main/java/net/minecraft/server/BlockCrops.java
+++ b/src/main/java/net/minecraft/server/BlockCrops.java
@@ -5,7 +5,7 @@ import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
 
 
 
-public class BlockCrops extends BlockPlant implements IBlockFragilePlantElement {
+public class BlockCrops extends BlockPlant implements IBlockFragilePlantElement, PandaGrowablePlant { // PandaSpigot
 
     public static final BlockStateInteger AGE = BlockStateInteger.of("age", 0, 7);
 
@@ -56,6 +56,12 @@ public class BlockCrops extends BlockPlant implements IBlockFragilePlantElement
         // CraftBukkit end
     }
 
+    // PandaSpigot start - expose shared plant growth state
+    public BlockStateInteger getGrowthState() {
+        return BlockCrops.AGE;
+    }
+    // PandaSpigot end
+
     protected static float getGrowthChance(Block block, World world, BlockPosition blockposition) {
         float f = 1.0F;
         BlockPosition blockposition1 = blockposition.down();
diff --git a/src/main/java/net/minecraft/server/BlockNetherWart.java b/src/main/java/net/minecraft/server/BlockNetherWart.java
index 4f7c81d3194fd4ce20bd7b6af9079f720c9962eb..09ca21cfa49b5d0ed8f2b64aff08fe7721137aa6 100644
--- a/src/main/java/net/minecraft/server/BlockNetherWart.java
+++ b/src/main/java/net/minecraft/server/BlockNetherWart.java
@@ -3,7 +3,7 @@ package net.minecraft.server;
 import java.util.Random;
 
 
-public class BlockNetherWart extends BlockPlant {
+public class BlockNetherWart extends BlockPlant implements PandaGrowablePlant { // PandaSpigot
 
     public static final BlockStateInteger AGE = BlockStateInteger.of("age", 0, 3);
 
@@ -71,6 +71,12 @@ public class BlockNetherWart extends BlockPlant {
         return ((Integer) iblockdata.get(BlockNetherWart.AGE)).intValue();
     }
 
+    // PandaSpigot start - expose shared plant growth state
+    public BlockStateInteger getGrowthState() {
+        return BlockNetherWart.AGE;
+    }
+    // PandaSpigot end
+
     protected BlockStateList getStateList() {
         return new BlockStateList(this, new IBlockState[] { BlockNetherWart.AGE});
     }
diff --git a/src/main/java/net/minecraft/server/BlockReed.java b/src/main/java/net/minecraft/server/BlockReed.java
index cd048b3dc1e87255bdcaa9742130a5643267b052..4cb565aeb92e06faaa7c1498d5a3cb7f6a5fd659 100644
--- a/src/main/java/net/minecraft/server/BlockReed.java
+++ b/src/main/java/net/minecraft/server/BlockReed.java
@@ -4,7 +4,7 @@ import java.util.Iterator;
 import java.util.Random;
 
 
-public class BlockReed extends Block {
+public class BlockReed extends Block implements PandaGrowablePlant { // PandaSpigot
 
     public static final BlockStateInteger AGE = BlockStateInteger.of("age", 0, 15);
 
@@ -117,6 +117,12 @@ public class BlockReed extends Block {
         return ((Integer) iblockdata.get(BlockReed.AGE)).intValue();
     }
 
+    // PandaSpigot start - expose shared plant growth state
+    public BlockStateInteger getGrowthState() {
+        return BlockReed.AGE;
+    }
+    // PandaSpigot end
+
     protected BlockStateList getStateList() {
         return new BlockStateList(this, new IBlockState[] { BlockReed.AGE});
     }
diff --git a/src/main/java/net/minecraft/server/BlockStateInteger.java b/src/main/java/net/minecraft/server/BlockStateInteger.java
index ab3b5165ccb609909a3b2526ec30af3749966690..206cc9e2b4fbb484e0e980e546dd57ea27cc8abe 100644
--- a/src/main/java/net/minecraft/server/BlockStateInteger.java
+++ b/src/main/java/net/minecraft/server/BlockStateInteger.java
@@ -9,6 +9,7 @@ import java.util.HashSet;
 public class BlockStateInteger extends BlockState<Integer> {
 
     private final ImmutableSet<Integer> allowedValues;
+    private final int maxValue; // PandaSpigot
 
     protected BlockStateInteger(String s, int i, int j) {
         super(s, Integer.class);
@@ -17,6 +18,7 @@ public class BlockStateInteger extends BlockState<Integer> {
         } else if (j <= i) {
             throw new IllegalArgumentException("Max value of " + s + " must be greater than min (" + i + ")");
         } else {
+            this.maxValue = j; // PandaSpigot
             HashSet hashset = Sets.newHashSet();
 
             for (int k = i; k <= j; ++k) {
@@ -32,6 +34,12 @@ public class BlockStateInteger extends BlockState<Integer> {
         return this.allowedValues;
     }
 
+    // PandaSpigot start - expose configured maximum value
+    public int getMaxValue() {
+        return this.maxValue;
+    }
+    // PandaSpigot end
+
     @Override
     public boolean equals(Object object) {
         if (this == object) {
diff --git a/src/main/java/net/minecraft/server/BlockStem.java b/src/main/java/net/minecraft/server/BlockStem.java
index ad1dea4c8e47452e1dc836403c5a9297b9ab3af1..2353b12e96eb2a5f1d4219d491671852734e994b 100644
--- a/src/main/java/net/minecraft/server/BlockStem.java
+++ b/src/main/java/net/minecraft/server/BlockStem.java
@@ -7,7 +7,7 @@ import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
 
 
 
-public class BlockStem extends BlockPlant implements IBlockFragilePlantElement {
+public class BlockStem extends BlockPlant implements IBlockFragilePlantElement, PandaGrowablePlant { // PandaSpigot
 
     public static final BlockStateInteger AGE = BlockStateInteger.of("age", 0, 7);
     public static final BlockStateDirection FACING = BlockStateDirection.of("facing", new Predicate() {
@@ -153,6 +153,12 @@ public class BlockStem extends BlockPlant implements IBlockFragilePlantElement {
         return ((Integer) iblockdata.get(BlockStem.AGE)).intValue();
     }
 
+    // PandaSpigot start - expose shared plant growth state
+    public BlockStateInteger getGrowthState() {
+        return BlockStem.AGE;
+    }
+    // PandaSpigot end
+
     protected BlockStateList getStateList() {
         return new BlockStateList(this, new IBlockState[] { BlockStem.AGE, BlockStem.FACING});
     }
diff --git a/src/main/java/net/minecraft/server/PandaGrowablePlant.java b/src/main/java/net/minecraft/server/PandaGrowablePlant.java
new file mode 100644
index 0000000000000000000000000000000000000000..69bbf0a4e38967c416557752d1acd0b0d189c36d
--- /dev/null
+++ b/src/main/java/net/minecraft/server/PandaGrowablePlant.java
@@ -0,0 +1,23 @@
+package net.minecraft.server;
+
+public interface PandaGrowablePlant {
+
+    static PandaGrowablePlant getById(int id) {
+        Block block = Block.getById(id);
+        return block instanceof PandaGrowablePlant ? (PandaGrowablePlant) block : null;
+    }
+
+    BlockStateInteger getGrowthState();
+
+    default int getGrowthLevel(IBlockData iblockdata) {
+        return iblockdata.get(this.getGrowthState()).intValue();
+    }
+
+    default int getMaxGrowthLevel() {
+        return this.getGrowthState().getMaxValue();
+    }
+
+    default boolean isFullyGrown(IBlockData iblockdata) {
+        return this.getGrowthLevel(iblockdata) >= this.getMaxGrowthLevel();
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
index 896940823f3719fae7adb9b3cb08f66021596a05..9124706d310f2156742b07dbd8d5b5a869f08af7 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -380,6 +380,38 @@ public class CraftBlock implements Block {
         return getBlockPower(BlockFace.SELF);
     }
 
+    // PandaSpigot start - Add growable plant API
+    private IBlockData getNMSBlockData() {
+        return chunk.getHandle().getBlockData(new BlockPosition(this.x, this.y, this.z));
+    }
+
+    private PandaGrowablePlant getGrowablePlant(IBlockData blockData) {
+        net.minecraft.server.Block block = blockData.getBlock();
+        return block instanceof PandaGrowablePlant ? (PandaGrowablePlant) block : null;
+    }
+
+    public boolean isGrowablePlant() {
+        return getGrowablePlant(getNMSBlockData()) != null;
+    }
+
+    public int getGrowthLevel() {
+        IBlockData blockData = getNMSBlockData();
+        PandaGrowablePlant plant = getGrowablePlant(blockData);
+        return plant != null ? plant.getGrowthLevel(blockData) : -1;
+    }
+
+    public int getMaxGrowthLevel() {
+        PandaGrowablePlant plant = getGrowablePlant(getNMSBlockData());
+        return plant != null ? plant.getMaxGrowthLevel() : -1;
+    }
+
+    public boolean isFullyGrown() {
+        IBlockData blockData = getNMSBlockData();
+        PandaGrowablePlant plant = getGrowablePlant(blockData);
+        return plant != null && plant.isFullyGrown(blockData);
+    }
+    // PandaSpigot end
+
     public boolean isEmpty() {
         return getType() == Material.AIR;
     }
